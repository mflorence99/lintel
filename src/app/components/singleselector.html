<ng-content select="[prefix]"></ng-content>

<select
  (change)="selectByIndex($event.target.selectedIndex)"
  [ngClass]="{placeholder: placeholder && (select.selectedIndex === 0)}"
  #select
  tabIndex=0
  autocomplete="off">

  <option
    [value]="null"
    *ngIf="placeholder">
    {{ placeholder }}
  </option>

  <option
    [attr.selected]="(value === option[0]) || ((ix === 0) && !placeholder)? true : null"
    [value]="option[0]"
    *ngFor="let option of options; let ix = index">
    {{ option[1] }}
  </option>

</select>

<ng-content select="[suffix]"></ng-content>
