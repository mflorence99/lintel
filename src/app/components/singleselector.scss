:host {
  align-items: center;
  background: var(--vscode-settings-textInputBackground);
  border: 1px solid transparent;
  color: var(--vscode-settings-textInputForeground);
  display: flex;
  height: 2rem;
  padding-right: 0.4rem;
}

:host(.focus) {
  border-color: var(--vscode-focusBorder);
}

select {
  background-color: var(--vscode-settings-textInputBackground);
  border: 0;
  color: var(--vscode-settings-textInputForeground);
  font-family: var(--vscode-font-family);
  font-size: var(--vscode-font-size);
  font-weight: var(--vscode-font-weight);
  height: 100%;
  width: 100%;
}

::ng-deep lintel-singleselector fa-icon {
  opacity: var(--lintel-opacity-icon);
}

::ng-deep lintel-singleselector fa-icon.disabled {
  opacity: var(--lintel-opacity-disabled);
}

::ng-deep lintel-singleselector [prefix] {
  margin-right: 0.5rem;
}

::ng-deep lintel-singleselector [suffix] {
  // NOTE: suffix is assumed to be a call to action
  cursor: pointer;
  margin-left: 0.5rem;
}
