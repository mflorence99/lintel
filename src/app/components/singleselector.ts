import { ChangeDetectionStrategy } from '@angular/core';
import { ChangeDetectorRef } from '@angular/core';
import { Component } from '@angular/core';
import { ControlValueAccessor } from '@angular/forms';
import { ElementRef } from '@angular/core';
import { Input } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { ViewChild } from '@angular/core';

import { forwardRef } from '@angular/core';

// NOTE: options can be one of the following:
// -- array of encoded values
// -- array of [encoded, decoded] values <-- PREFERRED

export type SingleselectorOptions =
  | SingleselectorValue[]
  | SingleselectorValue[][]
  | any[];

// NOTE: values can be one of the following:
// -- encoded value <-- PREFERRED

export type SingleselectorValue = number | string;

/**
 * Singleselect values via select/options
 *
 * NOTE: just enough to be able to match VSCode as well as possible
 *
 * @see https://blog.thoughtram.io/angular/2016/07/27/custom-form-controls-in-angular-2.html
 *
 * Reference is old, but still helpful
 */

@Component({
  changeDetection: ChangeDetectionStrategy.OnPush,
  providers: [
    {
      provide: NG_VALUE_ACCESSOR,
      useExisting: forwardRef(() => SingleselectorComponent),
      multi: true
    }
  ],
  selector: 'lintel-singleselector',
  templateUrl: 'singleselector.html',
  styleUrls: ['singleselector.scss']
})
export class SingleselectorComponent implements ControlValueAccessor {
  @Input() enabled = true;
  @Input() nameOfDecoded = 'value';
  @Input() nameOfEncoded = 'id';
  @Input() placeholder = '';

  @ViewChild('select', { static: true }) select: ElementRef;

  @Input()
  get options(): SingleselectorOptions {
    return this._options;
  }
  set options(options: SingleselectorOptions) {
    if (options && !this.origOptions) {
      this.origOptions = options;
      this._options = this.fromSingleselectorOptions(options);
    }
  }

  @Input()
  get value(): SingleselectorValue {
    return this._value;
  }
  set value(value: SingleselectorValue) {
    this._value = value;
    if (this.select?.nativeElement && this.select.nativeElement.options.length)
      this.select.nativeElement.selectedIndex = this.getOptionIndex(value);
    this.onChange?.(value);
    this.cdf.markForCheck();
  }

  // these shadow visible properties
  private _options: SingleselectorValue[][] = [];
  private _value: SingleselectorValue;

  private onChange: Function;
  private origOptions: SingleselectorOptions;

  /** ctor  */
  constructor(private cdf: ChangeDetectorRef) {}

  /** Get an indexed decoded option value */
  getOptionDecoded(ix: number): string {
    if (this.placeholder) ix = ix - 1;
    return ix >= 0 ? this.options[ix][1] : null;
  }

  /** Get an indexed encoded option value */
  getOptionEncoded(ix: number): string {
    if (this.placeholder) ix = ix - 1;
    return ix >= 0 ? this.options[ix][0] : null;
  }

  /** Get an indexed encoded option value */
  getOptionIndex(value: SingleselectorValue): number {
    const ix = this.options.findIndex((option) => option[0] === value);
    return this.placeholder ? ix + 1 : ix;
  }

  /** @see ControlValueAccessor */
  registerOnChange(fn): void {
    this.onChange = fn;
  }

  /** @see ControlValueAccessor */
  registerOnTouched(): void {}

  /** Select an an option by its index  */
  selectByIndex(ix: number): void {
    this.value = this.getOptionEncoded(ix);
  }

  /** @see ControlValueAccessor */
  writeValue(value): void {
    this.value = value;
  }

  // private methods

  private fromSingleselectorOptions(
    options: SingleselectorOptions
  ): SingleselectorValue[][] {
    let normalized: SingleselectorValue[][] = [];
    // NOTE: see above for different options for supplying options
    if (Array.isArray(options) && options.length > 0) {
      if (Array.isArray(options[0]))
        normalized = options as SingleselectorValue[][];
      else normalized = (options as string[]).map((option) => [option, option]);
    }
    return normalized;
  }
}
