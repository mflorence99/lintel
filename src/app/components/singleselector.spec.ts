import { SingleselectorComponent } from './singleselector';

import { prepare } from './component.spec';

import 'jest-extended';

import { ComponentFixture } from '@angular/core/testing';
import { TestBed } from '@angular/core/testing';

describe('SingleselectorComponent', () => {
  let component: SingleselectorComponent;
  let fixture: ComponentFixture<SingleselectorComponent>;

  beforeEach(() => {
    prepare();
    fixture = TestBed.createComponent(SingleselectorComponent);
    component = fixture.componentInstance;
  });

  test('registerOnChange', () => {
    component.registerOnChange(jest.fn());
    expect(component['onChange']).toBeTruthy();
  });

  test('registerOnTouched', () => {
    expect(component.registerOnTouched()).toBeFalsy();
  });

  test('writeValue', () => {
    expect(component.value).toBeUndefined();
    component.writeValue(42);
    expect(component.value).toBe(42);
  });

  test('number options, not decoded, no placeholder', () => {
    component.options = [1, 2, 3];
    expect(component.getOptionDecoded(1)).toBe(2);
    expect(component.getOptionDecoded(-1)).toBe(null);
    expect(component.getOptionEncoded(1)).toBe(2);
    expect(component.getOptionEncoded(-1)).toBe(null);
    expect(component.getOptionIndex(2)).toBe(1);
    component.selectByIndex(1);
    expect(component.value).toBe(2);
  });

  test('number options, decoded, no placeholder', () => {
    component.options = [
      [1, 'London'],
      [2, 'Paris'],
      [3, 'Rome']
    ];
    expect(component.getOptionDecoded(1)).toBe('Paris');
    expect(component.getOptionDecoded(-1)).toBe(null);
    expect(component.getOptionEncoded(1)).toBe(2);
    expect(component.getOptionEncoded(-1)).toBe(null);
    expect(component.getOptionIndex(2)).toBe(1);
    component.selectByIndex(1);
    expect(component.value).toBe(2);
  });

  test('number options, not decoded, with placeholder', () => {
    component.options = [1, 2, 3];
    component.placeholder = 'xxx';
    expect(component.getOptionDecoded(1)).toBe(1);
    expect(component.getOptionDecoded(0)).toBe(null);
    expect(component.getOptionEncoded(1)).toBe(1);
    expect(component.getOptionEncoded(0)).toBe(null);
    expect(component.getOptionIndex(2)).toBe(2);
    component.selectByIndex(1);
    expect(component.value).toBe(1);
  });

  test('string options, not decoded, no placeholder', () => {
    component.options = ['London', 'Paris', 'Rome'];
    expect(component.getOptionDecoded(1)).toBe('Paris');
    expect(component.getOptionDecoded(-1)).toBe(null);
    expect(component.getOptionEncoded(1)).toBe('Paris');
    expect(component.getOptionEncoded(-1)).toBe(null);
    expect(component.getOptionIndex('Paris')).toBe(1);
    component.selectByIndex(1);
    expect(component.value).toBe('Paris');
  });

  test('string options, decoded, no placeholder', () => {
    component.options = [
      ['l', 'London'],
      ['p', 'Paris'],
      ['r', 'Rome']
    ];
    expect(component.getOptionDecoded(1)).toBe('Paris');
    expect(component.getOptionDecoded(-1)).toBe(null);
    expect(component.getOptionEncoded(1)).toBe('p');
    expect(component.getOptionEncoded(-1)).toBe(null);
    expect(component.getOptionIndex('p')).toBe(1);
    component.selectByIndex(1);
    expect(component.value).toBe('p');
  });

  test('string options, not decoded, with placeholder', () => {
    component.options = ['London', 'Paris', 'Rome'];
    component.placeholder = 'xxx';
    expect(component.getOptionDecoded(1)).toBe('London');
    expect(component.getOptionDecoded(0)).toBe(null);
    expect(component.getOptionEncoded(1)).toBe('London');
    expect(component.getOptionEncoded(0)).toBe(null);
    expect(component.getOptionIndex('London')).toBe(1);
    component.selectByIndex(1);
    expect(component.value).toBe('London');
  });

  test('set number value', () => {
    component.options = [1, 2, 3];
    expect(component.value).toBeFalsy();
    component.registerOnChange((value) => expect(value).toBe(2));
    component.value = 2;
    expect(component.value).toBe(2);
  });

  test('set string value', () => {
    component.options = ['London', 'Paris', 'Rome'];
    expect(component.value).toBeFalsy();
    component.registerOnChange((value) => expect(value).toBe('Paris'));
    component.value = 'Paris';
    expect(component.value).toBe('Paris');
  });

  test('snapshot', () => {
    component.options = [
      ['l', 'London'],
      ['p', 'Paris'],
      ['r', 'Rome']
    ];
    component.placeholder = 'xxx';
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });
});
