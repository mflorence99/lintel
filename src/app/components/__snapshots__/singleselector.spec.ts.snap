// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SingleselectorComponent snapshot 1`] = `
<lintel-singleselector
  __ngContext__={[Function LRootView]}
  _options={[Function Array]}
  cdf={[Function ViewRef]}
  enabled={[Function Boolean]}
  nameOfDecoded={[Function String]}
  nameOfEncoded={[Function String]}
  origOptions={[Function Array]}
  placeholder={[Function String]}
  select={[Function ElementRef]}
>
  <select
    autocomplete="off"
    tabindex="0"
  >
    <option
      value="null"
    >
      xxx
    </option>
    
    <option
      value="l"
    >
       London 
    </option>
    <option
      value="p"
    >
       Paris 
    </option>
    <option
      value="r"
    >
       Rome 
    </option>
    
  </select>
</lintel-singleselector>
`;
